File: app/src/main/java/com/readrops/app/database/dao/AccountDao.java
package com.readrops.app.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;

import com.readrops.app.database.entities.account.Account;

import java.util.List;

@Dao
public abstract class AccountDao implements BaseDao<Account> {

    @Query("Select * from Account")
    public abstract LiveData<List<Account>> selectAll();

    @Query("Update Account set last_modified = :lastModified Where id = :accountId")
    public abstract void updateLastModified(int accountId, long lastModified);

    @Query("Update Account set current_account = 0 Where id Not In (:accountId)")
    public abstract void deselectOldCurrentAccount(int accountId);

    @Query("Update Account set current_account = 1 Where id = :accountId")
    public abstract void setCurrentAccount(int accountId);

    @Query("Select count(*) From Account Where account_type = :accountType")
    public abstract Integer getAccountCountByType(int accountType);

    @Query("Select count(*) From Account")
    public abstract Integer getAccountCount();

    @Query("Update Account set writeToken = :writeToken Where id = :accountId")
    public abstract void updateWriteToken(int accountId, String writeToken);
}

File: db/src/main/java/com/readrops/db/dao/AccountDao.kt
package com.readrops.db.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import com.readrops.db.entities.account.Account
import kotlinx.coroutines.flow.Flow

@Dao
interface AccountDao : BaseDao<Account> {

    override suspend fun insert(entity: Account): Long {
        val id = insertAccount(entity)
        updateCurrentAccount(id.toInt())
        return id
    }

    @Insert
    suspend fun insertAccount(entity: Account): Long

    @Query("Select * From Account")
    fun selectAllAccounts(): Flow<List<Account>>

    @Query("Select Count(*) From Account")
    suspend fun selectAccountCount(): Int

    @Query("Select * From Account Where current_account = 1")
    fun selectCurrentAccount(): Flow<Account?>

    @Query("Delete From Account")
    suspend fun deleteAllAccounts()

    @Query("Update Account set last_modified = :lastModified Where id = :accountId")
    suspend fun updateLastModified(lastModified: Long, accountId: Int)

    @Query("Update Account set notifications_enabled = :enabled Where id = :accountId")
    suspend fun updateNotificationState(accountId: Int, enabled: Boolean)

    @Query("Select notifications_enabled From Account Where id = :accountId")
    fun selectAccountNotificationsState(accountId: Int): Flow<Boolean>

    @Query("""Update Account set current_account = Case When id = :accountId Then 1 
        When id Is Not :accountId Then 0 End""")
    suspend fun updateCurrentAccount(accountId: Int)
}